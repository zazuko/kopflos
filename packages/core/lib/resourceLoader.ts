import type { AnyPointer, GraphPointer } from 'clownface'
import { isGraphPointer } from 'is-graph-pointer'
import type { NamedNode, Stream } from '@rdfjs/types'
import type { KopflosEnvironment } from './env/index.js'

export interface ResourceLoader {
  (iri: NamedNode, opts: {
    env: KopflosEnvironment
  }): Stream
}

export interface ResourceLoaderLookup {
  (resourceShape: GraphPointer, env: KopflosEnvironment): Promise<ResourceLoader | undefined>
}

export async function findResourceLoader(resourceShape: GraphPointer, env: KopflosEnvironment): Promise<ResourceLoader | undefined> {
  let resourceLoader: AnyPointer

  resourceLoader = resourceShape
    .out(env.ns.kopflos.resourceLoader)
    .out(env.ns.code.implementedBy)

  if (!isGraphPointer(resourceLoader)) {
    const api = resourceShape.in(env.ns.kopflos.api)
    resourceLoader = api
      .out(env.ns.kopflos.resourceLoader)
      .out(env.ns.code.implementedBy)

    if (!isGraphPointer(resourceLoader)) {
      resourceLoader = api
        .in(env.ns.kopflos.api)
        .out(env.ns.kopflos.resourceLoader)
        .out(env.ns.code.implementedBy)
    }
  }

  return env.load<ResourceLoader>(resourceLoader)
}

export const describe: ResourceLoader = (iri, { env }) => {
  return env.sparql.default.stream.query.construct(`DESCRIBE <${iri.value}>`)
}

export const fromOwnGraph: ResourceLoader = (iri, { env }) => {
  return env.sparql.default.stream.query.construct(`CONSTRUCT { ?s ?p ?o } WHERE { GRAPH <${iri.value}> { ?s ?p ?o } }`)
}
