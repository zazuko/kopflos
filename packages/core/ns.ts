import rdf from '@zazuko/env-node'

type Properties = 'api' | 'resourceLoader' | 'handler' | 'method' | 'config'
type Classes = 'Config' | 'Api' | 'ResourceShape' | 'Handler'
type Shorthands = 'DescribeLoader' | 'OwnGraphLoader'

export interface KopflosTerms extends Record<Properties | Classes | Shorthands, never>{
}

export const kl = rdf.namespace<keyof KopflosTerms>('https://kopflos.described.at/')
