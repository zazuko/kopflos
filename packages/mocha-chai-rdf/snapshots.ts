/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
