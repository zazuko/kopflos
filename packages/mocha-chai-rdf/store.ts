/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
