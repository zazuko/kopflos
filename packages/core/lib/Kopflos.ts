import type { IncomingHttpHeaders, OutgoingHttpHeaders } from 'node:http'
import type { DatasetCore, NamedNode, Stream } from '@rdfjs/types'
import type { GraphPointer, MultiPointer } from 'clownface'
import type { Options as EndpointOptions, StreamClient } from 'sparql-http-client/StreamClient.js'
import type { ParsingClient } from 'sparql-http-client/ParsingClient.js'
import { CONSTRUCT } from '@tpluscode/sparql-builder'
import { IN } from '@tpluscode/sparql-builder/expressions'
import type { KopflosEnvironment } from './env/index.js'
import { createEnv } from './env/index.js'
import type { ResourceShapeLookup, ResourceShapeMatch } from './resourceShape.js'
import defaultResourceShapeLookup from './resourceShape.js'
import { responseOr } from './responseOr.js'
import type { ResourceLoader, ResourceLoaderLookup } from './resourceLoader.js'
import { fromOwnGraph, findResourceLoader } from './resourceLoader.js'
import type { Handler, HandlerArgs, HandlerLookup } from './handler.js'
import { loadHandler } from './handler.js'
import type { HttpMethod } from './httpMethods.js'

interface KopflosRequest {
  iri: NamedNode
  method: HttpMethod
  headers: IncomingHttpHeaders
}

type ResultBody = Stream | DatasetCore | GraphPointer | Error
export interface ResultEnvelope {
  body?: ResultBody | string
  status?: number
  headers?: OutgoingHttpHeaders
}
export type KopflosResponse = ResultBody | ResultEnvelope

export interface Kopflos {
  get env(): KopflosEnvironment
  get apis(): MultiPointer
  handleRequest(req: KopflosRequest): Promise<ResultEnvelope>
}

interface Clients {
  stream: StreamClient
  parsed: ParsingClient
}

type Endpoint = string | EndpointOptions | Clients

export interface KopflosConfig {
  sparql: Record<string, Endpoint> & { default: Endpoint }
  codeBase?: string
}

export interface Options {
  dataset?: DatasetCore
  resourceShapeLookup?: ResourceShapeLookup
  resourceLoaderLookup?: ResourceLoaderLookup
  handlerLookup?: HandlerLookup
}

type Dataset = ReturnType<KopflosEnvironment['dataset']>

export default class Impl implements Kopflos {
  readonly dataset: Dataset
  readonly env: KopflosEnvironment

  constructor(private readonly config: KopflosConfig, private readonly options: Options = {}) {
    this.env = createEnv(config)

    this.dataset = this.env.dataset([
      ...options.dataset || [],
    ])
  }

  get graph() {
    return this.env.clownface({ dataset: this.dataset })
  }

  get apis(): MultiPointer {
    return this.graph.has(this.env.ns.rdf.type, this.env.ns.kopflos.Api)
  }

  async handleRequest(req: KopflosRequest): Promise<ResultEnvelope> {
    const result = await responseOr(this.findResourceShape(req.iri), (resourceShapeMatch: ResourceShapeMatch) => {
      const resourceShape = this.graph.node(resourceShapeMatch.resourceShape)

      return responseOr(this.findResourceLoader(resourceShape), loader => {
        const coreRepresentation = loader(req.iri, this)

        return responseOr(this.loadHandler(req.method, resourceShapeMatch, coreRepresentation), async handler => {
          const resourceGraph = this.env.clownface({
            dataset: await this.env.dataset().import(coreRepresentation),
          })
          const args: HandlerArgs = {
            resourceShape,
            env: this.env,
            subject: resourceGraph.node(resourceShapeMatch.subject),
            property: undefined,
            object: undefined,
          }
          if ('property' in resourceShapeMatch) {
            args.property = resourceShapeMatch.property
            args.object = resourceGraph.node(resourceShapeMatch.object)
          }

          return handler(args)
        })
      })
    })

    if (this.isEnvelope(result)) {
      return result
    }

    return {
      status: 200,
      body: result,
    }
  }

  async findResourceShape(iri: NamedNode): Promise<ResourceShapeMatch | KopflosResponse> {
    const resourceShapeLookup = this.options.resourceShapeLookup || defaultResourceShapeLookup

    return responseOr(resourceShapeLookup(iri, this), async candidates => {
      if (candidates.length === 0) {
        return { status: 404 }
      }

      if (candidates.length > 1) {
        return new Error('Multiple resource shapes found')
      }

      return candidates[0]
    })
  }

  async findResourceLoader(resourceShape: GraphPointer): Promise<ResourceLoader | KopflosResponse> {
    const resourceLoaderLookup = this.options.resourceLoaderLookup || findResourceLoader

    const loader = await resourceLoaderLookup(resourceShape, this.env)

    return loader || fromOwnGraph
  }

  async loadHandler(method: HttpMethod, resourceShapeMatch: ResourceShapeMatch, coreRepresentation: Stream): Promise<Handler | KopflosResponse> {
    const handlerLookup = this.options.handlerLookup || loadHandler

    const handler = await handlerLookup(resourceShapeMatch, method, this)

    if (handler) {
      return handler
    }

    if (!('property' in resourceShapeMatch) && (method === 'GET' || method === 'HEAD')) {
      return {
        status: 200,
        body: coreRepresentation,
      }
    }

    return { status: 405 }
  }

  private isEnvelope(arg: KopflosResponse): arg is ResultEnvelope {
    return 'body' in arg || 'status' in arg
  }

  static async fromGraphs(kopflos: Impl, ...graphs: Array<NamedNode | string>): Promise<void> {
    const graphsIris = graphs.map(graph => typeof graph === 'string' ? kopflos.env.namedNode(graph) : graph)
    const quads = CONSTRUCT`?s ?p ?o `
      .WHERE`
        GRAPH ?g {
          ?s ?p ?o
        }
        
        FILTER (?g ${IN(...graphsIris)})
      `.execute(kopflos.env.sparql.default.stream)

    for await (const quad of quads) {
      kopflos.dataset.add(quad)
    }
  }
}
