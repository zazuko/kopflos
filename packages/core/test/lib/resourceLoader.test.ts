import rdf, { Environment } from '@zazuko/env-node'
import { expect } from 'chai'
import { createStore } from 'mocha-chai-rdf/store.js'
import sinon from 'sinon'
import * as loader from '../../lib/resourceLoader.js'
import { ex } from '../../../testing-helpers/ns.js'
import type { KopflosConfig } from '../../lib/Kopflos.js'
import Kopflos from '../../lib/Kopflos.js'
import inMemoryClients from '../../../testing-helpers/in-memory-clients.js'
import 'mocha-chai-rdf/snapshots.js'
import { findResourceLoader } from '../../lib/resourceLoader.js'
import type { KopflosEnvironment } from '../../lib/env/index.js'
import { createEnv } from '../../lib/env/index.js'

describe('lib/resourceLoader', function () {
  function config(ctx: Mocha.Context): KopflosConfig {
    return {
      baseIri: 'http://example.com/',
      sparql: {
        default: inMemoryClients(ctx.rdf),
      },
    }
  }

  describe('built-in loaders', function () {
    let kopflos: Kopflos

    beforeEach(createStore(import.meta.url, { format: 'trig', loadAll: true, baseIri: ex }))

    beforeEach(async function () {
      kopflos = new Kopflos(config(this))
    })

    describe('describe', function () {
      it('should return a stream of a DESCRIBE query', async function () {
        // when
        const dataset = await rdf.dataset().import(loader.describe(ex.foo, kopflos))

        // then
        expect(dataset).canonical.toMatchSnapshot()
      })
    })

    describe('fromOwnGraph', function () {
      it('should contents of correct graph', async function () {
        // when
        const dataset = await rdf.dataset().import(loader.fromOwnGraph(ex.foo, kopflos))

        // then
        expect(dataset).canonical.toMatchSnapshot()
      })
    })
  })

  describe('findResourceLoader', function () {
    beforeEach(createStore(import.meta.url, { format: 'trig' }))

    class StubbedLoader {
      declare load: sinon.SinonStub

      init() {
        this.load = sinon.stub()
      }
    }

    let env: KopflosEnvironment

    beforeEach(function () {
      env = new Environment([StubbedLoader], { parent: createEnv(config(this)) })
    })

    context('when resource shape has a resource loader', function () {
      it('it will be used', async function () {
        // given
        const resourceShape = this.rdf.graph.node(ex.PersonShape)

        // when
        await findResourceLoader(resourceShape, env)

        // then
        expect(env.load).to.have.been.calledWith(sinon.match({
          term: ex.PersonResourceLoader,
        }))
      })
    })

    context('when resource shape has no resource loader', function () {
      it('API loader will be used', async function () {
        // given
        const resourceShape = this.rdf.graph.node(ex.PersonShape)

        // when
        await findResourceLoader(resourceShape, env)

        // then
        expect(env.load).to.have.been.calledWith(sinon.match({
          term: ex.ApiResourceLoader,
        }))
      })
    })

    context('when resource shape and API have no loaders', function () {
      it('loader from k:Config will be used', async function () {
        // given
        const resourceShape = this.rdf.graph.node(ex.PersonShape)

        // when
        await findResourceLoader(resourceShape, env)

        // then
        expect(env.load).to.have.been.calledWith(sinon.match({
          term: ex.ConfigResourceLoader,
        }))
      })
    })
  })
})
