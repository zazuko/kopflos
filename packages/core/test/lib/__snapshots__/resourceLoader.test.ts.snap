// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/resourceLoader built-in loaders describe should return a stream of a DESCRIBE query 1`] = `
"<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in default graph\\" .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph bar\\" .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph foo\\" .
"
`;

exports[`lib/resourceLoader built-in loaders fromOwnGraph should contents of correct graph 1`] = `
"<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph foo\\" .
"
`;

exports[`lib/resourceLoader describe should return a stream of a DESCRIBE query 1`] = `
"<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in default graph\\" .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph bar\\" .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph foo\\" .
"
`;

exports[`lib/resourceLoader fromOwnGraph should contents of correct graph 1`] = `
"<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#label> \\"in graph foo\\" .
"
`;
